### ISSUE TITLE
Missing character after expansion

### ISSUE THREAD

**Issue by bnt0** [co60ca/stella#3]
**Steps to reproduce:**
1. Installed the version `b151d1f` from the AUR.
2. Launch daemon with `stellad`
2. Set up configuration with `stella-config -Sak "lol" -v ":laughing out loud:" -e "true"`
3. When typing `lol` quickly in a text box (I tried XTerm and the Chromium address bar, and this very box :) sometimes the output is `:aughing out loud:` instead of the expected `:laughing out loud:`. Sometimes it takes a couple of times of quickly typing `lol` to reproduce the issue.

PS. really like your initiative, I'm getting familiar with the code base and the libraries at the moment. Let me know if there's something I can help with to get started.

**Comment by co60ca** [comment]
It's probably typing too fast. There is parameters to slow it down in Dispatcher.cc near the top
https://github.com/co60ca/stella/blob/master/src/Dispatcher.cc#L30. I had this issue at the beginning too, modifying the parameters helped most of the time.

What I'd like to do in the future is have a mode that uses the clipboard rather than using Xtest to type physical keys. (This would also allow us to write any bytes, not just ASCII)

I'm glad you tried it out, I'm sorry it not a perfect setup. The nice thing is most of the code is abstracted from the "typing" interface but you could always just copy & paste with xclip or something.

I plan on looking into xclip and seeing if it works for the project sometime, however you should not expect overly active development on this project.

**Comment by co60ca** [comment]
Other symptoms of this are erasing the first letter in the value text.

**Comment by co60ca** [comment]
Since we're GPL we can probably just borrow some of the code from xclip for this, the author made nice helper functions for dumping stuff in and out of X selections(clipboards)

**Comment by bnt0** [comment]
I agree, X selections would solve this, although I'm not sure why would XTest would fail at faster typing speed? 
We also want to avoid overwriting the user's clipboard though (maybe using [SECONDARY](https://wiki.archlinux.org/index.php/clipboard#Background) would be an easy way avoiding this). Alternatively, we can always just save the current clipboard content before usage and pasting it back after we're done.
Thanks for getting back to me btw, I'll have a look around too, and submit a couple more issues in the meanwhile :P

**Comment by co60ca** [comment]
In the past I've written a similar code base in Java, it suffered from many of the same problems. So this is just round two for me. I used the clipboard in the Java version as well, I just saved and restored the clipboard before and after. Secondary seems useful however. That would be the suggested route.

**Comment by bnt0** [comment]
Nice catch. If that's right, calling RobotCommon::typeString() event only after the user has released the last pressed key should solve this, I think.

**Comment by co60ca** [comment]
Right, I'm considering processing the key on key up rather than key down, another solution is to key up, key down, key up instead of key down, key up. Which will prevent any nonsense.

**Comment by co60ca** [comment]
Update: Tried a few things, doing up down up doesn't work, most likely because you're holding it down it cannot over ride that.

Second attempt, make the keys count on release not press. Which wasn't the case. Characters that can be typed (non control) can repeat, and will be captured by "typed" event. Released is still useful for control characters like backspace, however backspace can repeat and is not captured by "typed" events. Similar with tabs.

One idea is to keep track of how many keys are pressed by increment a count when we receive a "pressed" event and reduce when we have a "released" event. Then only allow the Robot to write when no keys are being pressed. But that's a bit hacky.

**Comment by co60ca** [comment]
Apparently if you type "fixes issue number" it instantly closes the issue. Please verify that this fix works suitably.

**Comment by bnt0** [comment]
I pulled your changes, and while it did fix the missing character issue, now insertions stop happening altogether after you press and hold a key. I compared with version 96abb76aca53633c23588a70c74e89fb53244063 and it was definitely introduced in the last two commits.
e.g. `lolllllllllll` *(press and hold `l`, then release)* `lol` *(no more expansions, for any keys)*

**Comment by co60ca** [comment]
I'm doing style right now, will check immediately after.

**Comment by co60ca** [comment]
Confirmed

**Comment by co60ca** [comment]
This is good enough for now, it will mess up if your holding more than one key on the last letter but will work for most cases

**Comment by bnt0** [comment]
OK, I'll move it to a separate issue then to keep track of it.

**Comment by bnt0** [comment]
Hi, I just tried it again, and think this issue (missing `l` in expansion) is back in 8689e61. Try typing `lol` and holding the `l` key.

**Comment by co60ca** [comment]
It will only type for me if you release the second `l` without letting it repeat. There for the missing l in the expansion doesn't occur for me since you're not holding the l down. Can you provide a more detailed example?

**Comment by bnt0** [comment]
It does insert after releasing the second `l`. When typing fast though I get `:aughing out loud:` or even `:aughing out loud::laughing out loud:` double insertions (and sometimes the expected `:laughing out loud:`) intermittently. 
It's hard to describe it in detail, as I couldn't yet find a way to reproduce it every time.

**Comment by co60ca** [comment]
As it stands if you are just hammering "lol" on your keyboard you are probably hitting an edge case that I might not be planning for. (Ideally it would be nice if it /always/ worked.) But the application is mostly meant to be used for posting a large section of text using a short snippet and waiting momentarily for the text to be printed to the screen.

If you cannot reproduce that way it may be out of scope.

**Comment by bnt0** [comment]
I can reproduce it, just not very reliably, as I don't know where/how the race condition happens. Generally speaking a relatively fast typer can run into this quite often, so I believe it's definitely within the scope, even for minimum functionality. Let's keep this issue open until the cause is found.
